test/storage/StorageTest.cc

```diff
@@ -0,0 +1,123 @@
+    assert(error_code == 0);  // Z_OK
+
+    // Start the namenode in a way that gives us a local pointer to zkWrapper.
+    unsigned short port = 5351;
+    nncli = new zkclient::ZkNnClient(zk);
+    nncli->register_watches();
+    nn_translator = new ClientNamenodeTranslator(port, nncli);
+  }
+
+  // Objects declared here can be used by all tests below.
+  zkclient::ZkNnClient *nncli;
+  ClientNamenodeTranslator *nn_translator;
+  std::shared_ptr<ZKWrapper> zk;
+};
+
+/**
+ * The following is an example of using StorageMetrics.
+ * This test puts a file into RDFS, then prints 2 metrics (SD and % space used)
+ */
+TEST_F(StorageTest, testExample) {
+  asio::io_service io_service;
+
+  RPCServer namenodeServer = nn_translator->getRPCServer();
+  std::thread(&RPCServer::serve, namenodeServer, std::ref(io_service))
+      .detach();
+
+  ASSERT_EQ(0, system("python "
+                          "/home/vagrant/rdfs/test/integration/generate_file.py"
+                          " > expected_""testfile1234"));
+  // Put a file into rdfs.
+  system(
+      "hdfs dfs -fs hdfs://localhost:5351 -D dfs.blocksize=1048576 "
+          "-copyFromLocal expected_testfile1234 /f");
+  sleep(5);
+
+  StorageMetrics metrics(NUM_DATANODES, zk);
+  LOG(INFO) << " ---- Standard Deviation of blocks per DataNode: " <<
+                                                 metrics.blocksPerDataNodeSD();
+  LOG(INFO) << " ---- Fraction of total space used: " <<
+                                                  metrics.usedSpaceFraction();
+}
+}  // namespace
+
+static inline void print_usage() {
+  LOG(ERROR) << "Storage test takes 0 or 1 arguments. \n"
+      "-h prints this message. \n"
+      "no argument runs the tests.";
+}
+
+static inline int runTests(int argc, char **argv) {
+  // Start up zookeeper
+  system("sudo /home/vagrant/zookeeper/bin/zkServer.sh stop");
+  system("sudo /home/vagrant/zookeeper/bin/zkServer.sh start");
+  sleep(10);
+  system("/home/vagrant/zookeeper/bin/zkCli.sh rmr /testing");
+  sleep(5);
+  system("rm -f expected_testfile1234 actual_testfile* temp* tfs*");
+
+  // initialize NUM_DATANODES datanodes
+  int32_t xferPort = 50010;
+  int32_t ipcPort = 50020;
+  for (int i = 0; i < NUM_DATANODES; i++) {
+    system(("truncate tfs" + std::to_string(i) + " -s 1000000000").c_str());
+    std::string dnCliArgs = "-x " +
+        std::to_string(xferPort + i) + " -p " + std::to_string(ipcPort + i)
+        + " -b tfs" + std::to_string(i) + " &";
+    std::string cmdLine =
+        "bash -c \"exec -a StorageTestServer" + std::to_string(i) +
+            " /home/vagrant/rdfs/build/rice-datanode/datanode " +
+            dnCliArgs + "\" & ";
+    system(cmdLine.c_str());
+    sleep(3);
+  }
+
+  // Initialize and run the tests
+  ::testing::InitGoogleTest(&argc, argv);
+  int res = RUN_ALL_TESTS();
+
+  // Kill the DataNodes and NameNodes, and stop zookeeper
+  system("pkill -f StorageTestServer*");
```

Comment:
If you run this test with other namenodes running on the machine, what happens? Seems like it kills everything.

Reply:
the line after this kills any process matching "namenode" so i think it would kill any namenode running on the machine. in the context of test setup/teardown i think that's ok
line 107 only kills datanodes started in this test though, since other tests use different names.